source/_drafts/Feature-Spotlight-Middleware.md

```diff
@@ -0,0 +1,187 @@
+The page can be rendered with a specific catalog ID and a special MIME type, like so:
+```html
+<video controls class="video-js">
+  <source src="123" type="video/my-catalog">
+</video>
+```
+Then, you could register a middleware for that route -- the type `video/my-catalog`.
+```js
+// middleware methods get the player instance as an argument
+videojs.use('video/my-catalog', function(player) {
+
+  // middleware are expected to return an object with the methods on it.
+  // It can be a plain object or an instance of something.
+  return {
+
+    // setSource allows you to tell Video.js whether you're going to be handling the source or not
+    setSource(srcObj, next) {
+      const id = srcObj.src;
+
+      videojs.xhr({
+        uri: '/getVideo?id=' + id
+      }, function(err, res, body) {
+
+        // pass null as the first argument to say everything is going fine and we can handle it.
+        next(null, {
+          src: body.sourceUrl,
+          type: body.sourceType
+        })
+      });
+    }
+  };
+});
+```
+Then, when Video.js initializes, it'll go through and call the middleware that are set for `video/my-catalog`.
+
+Server Side Ad Insertion (SSAI) is a great fit for middleware. It showcases the ability to intercept the play and tech interactions.
+For example, you have both ads and content in your HLS playlist but you want the controls to display different times for each.
+When the content video is playing, you want to display a timeline of 5 minutes. And for ads you want to display 30 seconds for the pre-roll.
+Yet, currently, the timeline is showing a combined duration of 5 minutes and 30 seconds.
+So, you can add a middleware that knows when the ad is being played and tells the player that the duration is 30 seconds
+and when the content is playing that the duration is 5 minutes.
+```js
+// register a star-middleware because HLS has two mimetypes
+videojs.use('*', function(player) {
+  return {
+    setSource(srcObj, next) {
+      const type = srcObj.type;
+
+      if (type !== 'application/x-mpegurl' && type !== 'application/vnd.apple.mpegurl') {
+
+        // call next with an error to signal you cannot handle the source
+        next(new Error('Source is not an HLS source'));
+      } else {
+
+        // in here we know we're playing back an HLS source.
+        // We don't want to do anything special for it, so, pass along the source along with a null.
+        next(null, srcObj);
+      }
+    },
+
+    // this method gets called on the tech and then up the middleware chain providing the values as you go along
+    duration(durationFromTech) {
+      if (areWeCurrentlyPlayingAnAd(durationFromTech)) {
+        // since we're now in an ad, return the ad duration
+        return 30;
+      } else {
+        // we're playing back content, so, return that duration
+        return 5 * 60;
+      }
+    }
+  }
+});
+```
+
+A simple but interesting middleware to look at is the [playbackrate-adjuster][pra].
```

Comment:
A final heading here for this would be good, perhaps: "The `playbackrate-adjuster` an implementation in the wild"